services/prometheus/scrape_configs.go

```diff
@@ -97,18 +132,200 @@ func (svc *Service) putToConsul(scs []ScrapeConfig) error {
 	return svc.consul.PutKV(ConsulKey, b)
 }
 
+// getTargetsHealth gets all targets from Prometheus and converts response to job -> instance -> health map.
+func (svc *Service) getTargetsHealth(ctx context.Context) (map[string]map[string]Health, error) {
+	u := *svc.baseURL
+	u.Path = path.Join(u.Path, "api/v1/targets")
+	resp, err := svc.client.Get(u.String())
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	defer resp.Body.Close()
+
+	// Copied from vendor/github.com/prometheus/prometheus/web/api/v1/api.go to avoid a ton of dependencies
+	type target struct {
+		DiscoveredLabels model.LabelSet `json:"discoveredLabels"`
+		Labels           model.LabelSet `json:"labels"`
+		ScrapeURL        string         `json:"scrapeUrl"`
+		LastError        string         `json:"lastError"`
+		LastScrape       time.Time      `json:"lastScrape"`
+		Health           string         `json:"health"` // 	"unknown", "up", or "down"
+	}
+	type result struct {
+		Status string `json:"status"`
+		Data   struct {
+			ActiveTargets []target `json:"activeTargets"`
+		} `json:"data"`
+	}
+	var res result
+	if err = json.NewDecoder(resp.Body).Decode(&res); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	health := make(map[string]map[string]Health)
+	for _, target := range res.Data.ActiveTargets {
+		job := string(target.Labels["job"])
+		instance := string(target.Labels["instance"])
+		if health[job] == nil {
+			health[job] = make(map[string]Health)
+		}
+
+		health[job][instance] = HealthUnknown
+		switch target.Health {
+		case "down":
+			health[job][instance] = HealthDown
+		case "up":
+			health[job][instance] = HealthUp
+		}
+	}
+	return health, nil
+}
+
+// checkReachability checks that given targets can be reached from PMM Server.
+// reachabilityCh is closed when this method returns.
+func (svc *Service) checkReachability(ctx context.Context, cfg *ScrapeConfig, targets []string, reachabilityCh chan<- ScrapeTargetReachability) {
+	// use ephemeral transport with small timeouts and disabled HTTP keep-alive
+	transport := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   5 * time.Second,
+			DualStack: true,
+		}).DialContext,
+		TLSHandshakeTimeout: 5 * time.Second,
```

Comment:
The lower timeouts don't work?

Reply:
They should, I'm just being too cautious there.

Reply:
You can always try lower first. Especially it's a health-check. I would also add it as global, easy to modify and avoids duplicate timeout definitions.

Reply:
But it is not duplicated anywhere.